### ISSUE TITLE
v1.12.6: containers running inside a single user-defined network are sometimes unable to talk to each other

### ISSUE THREAD

**Issue by pkubatrh** [moby/moby#30312]
**Description**

While running test cases for some of our Docker images I have hit a curious issue, that seems to be connected to user-defined docker networks.
The test cases, which are run in parallel, are each creating their own docker network while using the --subnet (and --ip) argument in order to be able to use static IPs (as this is needed during the tests). This works most of the time. However, once in a while one of the tests starts taking too long to finish as it gets stuck while setting up a new container that is trying to contact another container in the same network, but does not receive a response.

**Steps to reproduce the issue:**
The issue is hard to reproduce as it manifests itself randomly while running test cases for docker images long enough. The script that is able to run the tests can be located at:
https://github.com/pkubatrh/postgresql-container/blob/priv_net_reproducer/reproducer.sh
The test cases themselves can be found over here:
https://github.com/pkubatrh/postgresql-container/blob/priv_net_reproducer/hack/run_test#L468

1. Run the reproducer until you see a test cycle is taking too long to complete
2. Exec into the container that is having troubles
3. Try to ping any other container located inside the same network

**Describe the results you received:**
The other container does not respond to the ping

**Describe the results you expected:**
The other container should be able to respond to the ping

**Additional information you deem important (e.g. issue happens only occasionally):**
Any additional containers created manually inside the problematic network are also not able to contact anyone else but themselves (and the network's gateway).
The issue is present even after removing the problematic network and recreating it (with the same --subnet value). Restarting the docker daemon did not help to recover from the issue either.

**Output of `docker version`:**
Client:
 Version:         1.12.6
 API version:     1.24
 Package version: docker-common-1.12.6-4.gitf499e8b.fc25.x86_64
 Go version:      go1.7.4
 Git commit:      f499e8b/1.12.6
 Built:           Fri Jan 13 11:03:22 2017
 OS/Arch:         linux/amd64

Server:
 Version:         1.12.6
 API version:     1.24
 Package version: docker-common-1.12.6-4.gitf499e8b.fc25.x86_64
 Go version:      go1.7.4
 Git commit:      f499e8b/1.12.6
 Built:           Fri Jan 13 11:03:22 2017
 OS/Arch:         linux/amd64

**Output of `docker info`:**
Containers: 9
 Running: 1
 Paused: 0
 Stopped: 8
Images: 64
Server Version: 1.12.6
Storage Driver: devicemapper
 Pool Name: docker-253:1-131451-pool
 Pool Blocksize: 65.54 kB
 Base Device Size: 10.74 GB
 Backing Filesystem: xfs
 Data file: /dev/loop0
 Metadata file: /dev/loop1
 Data Space Used: 862.1 MB
 Data Space Total: 107.4 GB
 Data Space Available: 4.921 GB
 Metadata Space Used: 1.937 MB
 Metadata Space Total: 2.147 GB
 Metadata Space Available: 2.146 GB
 Thin Pool Minimum Free Space: 10.74 GB
 Udev Sync Supported: true
 Deferred Removal Enabled: false
 Deferred Deletion Enabled: false
 Deferred Deleted Device Count: 0
 Data loop file: /var/lib/docker/devicemapper/devicemapper/data
 WARNING: Usage of loopback devices is strongly discouraged for production use. Use `--storage-opt dm.thinpooldev` to specify a custom block storage device.
 Metadata loop file: /var/lib/docker/devicemapper/devicemapper/metadata
 Library Version: 1.02.136 (2016-11-05)
Logging Driver: journald
Cgroup Driver: systemd
Plugins:
 Volume: local
 Network: null host bridge overlay
Swarm: inactive
Runtimes: oci runc
Default Runtime: oci
Security Options: seccomp selinux
Kernel Version: 4.9.3-200.fc25.x86_64
Operating System: Fedora 25 (Twenty Five)
OSType: linux
Architecture: x86_64
Number of Docker Hooks: 2
CPUs: 2
Total Memory: 3.86 GiB
Name: host-172-16-36-167
ID: DILD:GVRX:IXIC:DAZS:3ZHY:MSXG:IOCX:4YZU:UTMM:GDBN:KQ6K:GEWH
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Insecure Registries:
 127.0.0.0/8
Registries: docker.io (secure)

**Additional environment details (AWS, VirtualBox, physical, etc.):**

Running on a Fedora 25 machine inside OpenStack

**Comment by pkubatrh** [comment]
This seems to be happening because responses to pings (or any other form of communication over IP really) for some reason get routed to the network's gateway and do not find their way back to the container that issued the ping. Unfortunately I am no longer able to reproduce, as I am now running into a very similar issue, which I believe has the same underlying problem:

When the new issue is reproduced, the containers are able to communicate with each other (this is the major difference between those two issues). In the case of a ping, all seems to be ok when trying to look from the sending container's view:

`08:49:52.424352 02:42:ac:12:04:04 > 02:42:ac:12:04:03, ethertype IPv4 (0x0800), length 98: 172.18.4.4 > 172.18.4.3: ICMP echo request, id 655, seq 1, length 64
08:49:52.424434 02:42:ac:12:04:03 > 02:42:ac:12:04:04, ethertype IPv4 (0x0800), length 98: 172.18.4.3 > 172.18.4.4: ICMP echo reply, id 655, seq 1, length 64`

However, the responding container sees the incoming packet as being sent from the network's gateway's IP (but the source MAC field corresponds to the sender container, not the gateway) and responds to the same IP (the MAC is set to gateway's address here, as expected):

`08:58:34.652942 02:42:ac:12:04:04 > 02:42:ac:12:04:03, ethertype IPv4 (0x0800), length 98: 172.18.4.1 > 172.18.4.3: ICMP echo request, id 657, seq 2, length 64
08:58:34.652962 02:42:ac:12:04:03 > 02:42:70:09:c9:b5, ethertype IPv4 (0x0800), length 98: 172.18.4.3 > 172.18.4.1: ICMP echo reply, id 657, seq 2, length 64`

Exactly the same problem arises when the sender and responder are swapped so it seems not to be linked to a single container but to the network itself:

`09:01:05.500914 02:42:ac:12:04:03 > 02:42:ac:12:04:04, ethertype IPv4 (0x0800), length 98: 172.18.4.3 > 172.18.4.4: ICMP echo request, id 177, seq 5, length 64
09:01:05.500973 02:42:ac:12:04:04 > 02:42:ac:12:04:03, ethertype IPv4 (0x0800), length 98: 172.18.4.4 > 172.18.4.3: ICMP echo reply, id 177, seq 5, length 64
09:01:05.500939 02:42:ac:12:04:03 > 02:42:ac:12:04:04, ethertype IPv4 (0x0800), length 98: 172.18.4.1 > 172.18.4.4: ICMP echo request, id 177, seq 5, length 64
09:01:05.500956 02:42:ac:12:04:04 > 02:42:70:09:c9:b5, ethertype IPv4 (0x0800), length 98: 172.18.4.4 > 172.18.4.1: ICMP echo reply, id 177, seq 5, length 64`

Routing is set up correctly on all containers involved and arp cache has the correct values, so the problem should not be connected to either of those. We have tried limiting access to the network creation using `flock` (since the issue has so far been only reproduced when the tests are run in parallel) but had no luck so far.

**Comment by thaJeztah** [comment]
Let me close this ticket for now, as it looks like it went stale.
